File: app/src/main/kotlin/tech/mmarca/openvitals/domain/model/DashboardQuery.kt
package tech.mmarca.openvitals.domain.model

import java.time.LocalDate
import tech.mmarca.openvitals.domain.model.RefreshMode
import tech.mmarca.openvitals.domain.preferences.ActivityWeekMode
import tech.mmarca.openvitals.domain.preferences.SleepRangeMode

data class DashboardQuery(
    val date: LocalDate = LocalDate.now(),
    val sleepRangeMode: SleepRangeMode = SleepRangeMode.EVENING_18H,
    val activityWeekMode: ActivityWeekMode = ActivityWeekMode.MONDAY_TO_SUNDAY,
    val visibleMetrics: Set<DashboardMetric> = DashboardMetric.entries.toSet(),
    val refreshMode: RefreshMode = RefreshMode.NORMAL,
    val includeHistoricalBaselines: Boolean = true,
    val includeWeeklyTrainingSignals: Boolean = true,
)

enum class DashboardMetric {
    STEPS,
    DISTANCE,
    CALORIES_OUT,
    ACTIVE_CALORIES,
    FLOORS,
    ELEVATION,
    WHEELCHAIR_PUSHES,
    WORKOUT,
    SLEEP,
    HYDRATION,
    CALORIES_IN,
    PROTEIN,
    CARBS,
    FAT,
    CAFFEINE,
    WEIGHT,
    HEIGHT,
    BMI,
    FFMI,
    BODY_FAT,
    LEAN_MASS,
    BMR,
    BONE_MASS,
    BODY_WATER_MASS,
    AVG_HEART_RATE,
    RESTING_HEART_RATE,
    HRV,
    BLOOD_PRESSURE,
    SPO2,
    VO2_MAX,
    RESPIRATORY_RATE,
    BODY_TEMPERATURE,
    BLOOD_GLUCOSE,
    SKIN_TEMPERATURE,
    WEEKLY_CARDIO_LOAD,
    INTENSITY_MINUTES,
    MINDFULNESS,
    CYCLE,
}

File: lib/domain/model/dashboard_query.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../core/time/local_date.dart';
import '../preferences/activity_week_mode.dart';
import '../preferences/sleep_range_mode.dart';
import 'refresh_mode.dart';

part 'dashboard_query.freezed.dart';

@freezed
abstract class DashboardQuery with _$DashboardQuery {
  const factory DashboardQuery.build({
    required LocalDate date,
    required SleepRangeMode sleepRangeMode,
    required ActivityWeekMode activityWeekMode,
    required Set<DashboardMetric> visibleMetrics,
    required RefreshMode refreshMode,
    required bool includeHistoricalBaselines,
    required bool includeWeeklyTrainingSignals,
  }) = _DashboardQuery;

  factory DashboardQuery({
    LocalDate? date,
    SleepRangeMode sleepRangeMode = SleepRangeMode.evening18h,
    ActivityWeekMode activityWeekMode = ActivityWeekMode.mondayToSunday,
    Set<DashboardMetric>? visibleMetrics,
    RefreshMode refreshMode = RefreshMode.normal,
    bool includeHistoricalBaselines = true,
    bool includeWeeklyTrainingSignals = true,
  }) =>
      DashboardQuery.build(
        date: date ?? LocalDate.now(),
        sleepRangeMode: sleepRangeMode,
        activityWeekMode: activityWeekMode,
        visibleMetrics: visibleMetrics ?? DashboardMetric.values.toSet(),
        refreshMode: refreshMode,
        includeHistoricalBaselines: includeHistoricalBaselines,
        includeWeeklyTrainingSignals: includeWeeklyTrainingSignals,
      );
}

enum DashboardMetric {
  steps('STEPS'),
  distance('DISTANCE'),
  caloriesOut('CALORIES_OUT'),
  activeCalories('ACTIVE_CALORIES'),
  floors('FLOORS'),
  elevation('ELEVATION'),
  wheelchairPushes('WHEELCHAIR_PUSHES'),
  workout('WORKOUT'),
  sleep('SLEEP'),
  hydration('HYDRATION'),
  caloriesIn('CALORIES_IN'),
  protein('PROTEIN'),
  carbs('CARBS'),
  fat('FAT'),
  caffeine('CAFFEINE'),
  weight('WEIGHT'),
  height('HEIGHT'),
  bmi('BMI'),
  ffmi('FFMI'),
  bodyFat('BODY_FAT'),
  leanMass('LEAN_MASS'),
  bmr('BMR'),
  boneMass('BONE_MASS'),
  bodyWaterMass('BODY_WATER_MASS'),
  avgHeartRate('AVG_HEART_RATE'),
  bodyEnergy('BODY_ENERGY'),
  restingHeartRate('RESTING_HEART_RATE'),
  hrv('HRV'),
  bloodPressure('BLOOD_PRESSURE'),
  spo2('SPO2'),
  vo2Max('VO2_MAX'),
  respiratoryRate('RESPIRATORY_RATE'),
  bodyTemperature('BODY_TEMPERATURE'),
  bloodGlucose('BLOOD_GLUCOSE'),
  skinTemperature('SKIN_TEMPERATURE'),
  weeklyCardioLoad('WEEKLY_CARDIO_LOAD'),
  intensityMinutes('INTENSITY_MINUTES'),
  mindfulness('MINDFULNESS'),
  cycle('CYCLE');

  const DashboardMetric(this.storageName);

  /// Original Kotlin `.name` used for persistence round-trips.
  final String storageName;

  static DashboardMetric? fromStorage(String value) {
    for (final entry in values) {
      if (entry.storageName == value) return entry;
    }
    return null;
  }
}
